lms/www/batch/learn.py

```diff
@@ -105,12 +105,63 @@ def get_page_extensions(context):
 		e.set_context(context)
 	return extensions
 
+class LessonBookmark:
+	"""
+	This represents a simple data structure to represent a lesson bookmark.
+	"""
+	def __init__(self, lesson_number: str) -> None:
+		self.__test_param_or_raise_exception(lesson_number)
+		_lesson_number = f"{lesson_number}." if not "." in lesson_number else lesson_number
+		first, second = _lesson_number.split(".")
+
+		self.__value = (int(first), int(second) or 0)	# second would be "" if `lesson_number` is something like "7"
+
+	def __eq__(self, other: object) -> bool:
+		if isinstance(other, self.__class__):
+			return self.__value == other.value
+		return NotImplemented
+	
+	def __gt__(self, other: object) -> bool:
+		if isinstance(other, self.__class__):
+			return self.__value > other.value
+		return False
+
+	def __repr__(self) -> str:
+		return f"{self.__value[0]}.{self.__value[1]}"
+	
+	def __test_param_or_raise_exception(self, string: str) -> None:
+		"""
+		Tests that a given string is in the format "n" or "n.n" where n is a numeric
+		character. If the string does not match the expected format, `TypeError` is
+		raised.
+		"""
+		import re
+		import sys
+		expected_format = r'^\d+\.?\d*$'
+
+		try:
+			if not re.match(expected_format, string):
+				raise TypeError("""Expected a 'str' in the format 'n' or 'n.n' where n 
+				   is a numeric character. Example: '7' or '7.10""")
+		except TypeError as e:
+			tb = sys.exc_info()[2]
+			raise TypeError("""Expected a 'str' in the format 'n' or 'n.n' where n 
+				   is a numeric character. Example: '7' or '7.10""").with_traceback(tb)
+	
+	@property
+	def value(self):
+		return self.__value
+	
+	@property
+	def readable_value(self):
+		return self.__repr__()
+
 
 def get_neighbours(current, lessons):
-	current = flt(current)
```

Comment:
@tundebabzy only removing this line is enough. What is the need of this bookmark class? According to me this file online needs this line removed and no other change.

Reply:
@pateljannat, do you mean removing the cast to float using `flt`? If yes, that doesn't solve the issue with how python sorts strings. Sorting `["1.1", "2.1", "12.1"]` will result in `["1.1", "12.1", "2.1"]`.

The proper way is actually to store the "bookmark" as a tuple of `int`s like `[(1,1),(1,12), (12,1)]` which python can sort as expected

The `Bookmark` class is a data structure to abstract this and make the code more readable.

Reply:
> that doesn't solve the issue with how python sorts strings

You are right, this won't solve the problem. I think your approach to converting it into a tuple will fix the issue. 

> make the code more readable.

I'm afraid I have to disagree with this part. Converting the numbers array into tuples and sorting them can be done in fewer lines of code.

```
def get_neighbours(current, lessons):
	numbers = [lesson.number for lesson in lessons]
	tuples_list = [tuple(map(int, s.split('.'))) for s in numbers]
	sorted_tuples = sorted(tuples_list)
	sorted_numbers = ['.'.join(map(str, t)) for t in sorted_tuples]
	index = sorted_numbers.index(current)

	return {
		"prev": sorted_numbers[index - 1] if index - 1 >= 0 else None,
		"next": sorted_numbers[index + 1] if index + 1 < len(sorted_numbers) else None,
	}
```

Reply:
Fewer lines in not necessarily more readable.

I was actually more concerned with the overhead and performance penalty of the class abstraction since its like 50% slower than without the abstraction even though negligible for real life scenarios. I will remove the abstraction as requested